bitmovin/bitmovin-player-web-integrations-yospace
# MECBM-659: Duplicate Default Impressions and VPAID Stutter on Pre-Roll

## Overview

This PR was a joint effort with @mcarriga.

This fix handles two of the issues that came out of [MECBM-659](https://jiraprod.turner.com/browse/MECBM-659).
* Duplicate Default Impressions for VPAID ads
* VPAID pre-rolls stuttering on start

## Resolution

### Duplicate Impressions

For duplicate default impressions, it appears that for an earlier version of the Yospace Client SDK (`1.7.14`) that the default impressions were not being reported correctly, so the Bitmovin player was keeping `<Impression>` tags in the VAST response, which would in turn ensure they are fired. As of the Yospace Client SDK `1.8.1` this issue is no longer present. So the following block was once again firing `defaultimpression` beacons, alongside the core Bitmovin player.

### Stuttering

@mcarriga Implemented a fix here that uses module events which return us the earliest instance of when Yospace ad break data is available. Once we have that data, we prematurely schedule the VPAID ad break, which intercepts the stitched in Yospace ad break earlier than before, which prevents the stutter from occurring.

## Testing

* Ensure you're using Node v10 or v11 for Gulp to run
* Ensure you also have access to the [TUB Lib repo and can copy the Yospace Client SDK](https://github.com/TurnerOpenPlatform/tub-lib/blob/develop/web/scripts/yo-ad-management.min.js) JS into the local `yospace` directory in this repo
* `npm install && npm start` should open a browser window
* Run the demo with the following query parameters (https://localhost:9000/?aggressiveVpaid=true&vpaid=true):
  * `aggressiveVpaid=true` will enable the aggressive pre-rendering approach (which is enabled by default)
  * `vpaid=true` is the previous flag which allows for toggling VPAID on and off as a whole
* Select the `VOD Preroll VPAID` asset
  * Observe that the stutter is no longer present across all browsers
  * Observe under the network traffic the following beacons fire without duplicates, for the `bea4` freewheel endpoint
    * `slotimpression` 
    * `defaultimpression` (this was the duplicate offender)
    * `firstquartile`
    * `midpoint`
    * `thirquartile`
    * `complete`
* Select the `VOD Midroll VPAID` asset, and perform the same validation as above, seeking into the VOD `3:20` into playback which will trigger a Midroll ad break around `3:30`.
